File: Medium Question/198. House Robber.cpp
class Solution {
public:
 int solve(vector<int>&nums)
{
    int n=nums.size();
     int prev=nums[0];
  int prev2=0;
    for(int i=1;i<n;i++)
    {
        int take = nums[i];
        if(i>1)
            take += prev2;

        int not_take = prev;

        int cur = max(take, not_take);
        prev2 = prev;
        prev = cur;
    }

    return prev;
}
    int rob(vector<int>& nums) {
         int n=nums.size();

if(n==1) return nums[0];

    return solve(nums);
    }
};

File: Medium Question/198.House Robber.java
class Solution {
    public int rob(int[] nums) {
        int n=nums.length;
//tabulation method
// 		int[] dp= new int[n];
//         dp[0]=nums[0];
// int neg=0;
// for(int i=1;i<n;i++){
	
// 	int take=nums[i];
// 	if(i>1){
// 		take=take+dp[i-2];
// 	}
// 	int notpick=0+dp[i-1];
// 	dp[i]=Math.max(take,notpick);

// }
//  return dp[n-1];}}


//space optimization
int prev=nums[0];
int prev2=0;
for(int i=1;i<n;i++){
	int pick=nums[i]+prev2;
	int notpick=0+prev;
	int curr=Math.max(pick,notpick);
	prev2=prev;
	prev=curr;


	
}
return prev;}}
